include/kernel/Kernel.h

```diff
@@ -129,8 +132,67 @@ class Kernel
     ScratchViews<DoubleType> & /* elemScratchViews */,
     int /* elemFaceOrdinal */)
   {}
+
+  /** Execute for NGP Kernels
+   *
+   */
+  KOKKOS_FUNCTION
+  virtual void execute(
+    SharedMemView<DoubleType**, DeviceShmem>&,
+    SharedMemView<DoubleType*, DeviceShmem>&,
+    ScratchViews<double, DeviceTeamHandleType, DeviceShmem>&)
+  {}
+};
+
+/** Kernel that can be transferred to a device
+ *
+ */
+template<typename T>
+class NGPKernel : public Kernel
+{
+public:
+  KOKKOS_FORCEINLINE_FUNCTION
+  NGPKernel() = default;
+
+  virtual ~NGPKernel()
+  {
+    if (deviceCopy_ != nullptr)
+      kokkos_free_on_device(deviceCopy_);
+  }
+
+  virtual Kernel* create_on_device() final
+  {
+    if (deviceCopy_ != nullptr)
+      kokkos_free_on_device(deviceCopy_);
+    deviceCopy_ = create_device_expression<T>(*dynamic_cast<T*>(this));
+    return deviceCopy_;
+  }
+
+  T* device_copy() const { return deviceCopy_; }
+
+protected:
+  T* deviceCopy_{nullptr};
 };
 
+/** Wrapper object to hold pointers to kernel instances within a Kokkos::View
+ *
+ */
+struct NGPKernelInfo
+{
+  KOKKOS_FUNCTION
+  NGPKernelInfo() = default;
+
+  NGPKernelInfo(Kernel& kernel)
+  {
+    kernel_ = kernel.create_on_device();
```

Comment:
kernel.create_on_devce() always returns "this".  Why not just kernel_(&kernel).

Reply:
@overfelt `NGPKernel::create_on_device` returns the device pointer and not `this`. I have a dummy function in the base `Kernel` class for migration purposes so that the code will compile for non-NGP ready kernels. `Kernel::create_on_device` should be a pure virtual function eventually. 

Although you are correct, I can do in the initializer list. 

```
kernel_(kernel.create_on_device())
```